tests/run/double-pattern-type.scala

```diff
@@ -0,0 +1,40 @@
+case class C1(i: String, s: Int) { def isEmpty = false; def get = ("EMPTY", -1) }
+case class C2(i: String, s: String) { def isEmpty = false; def get = (-1, -2, -3) }
+
+object Test {
+  def main(args: Array[String]): Unit = {
+    // When both Product and name based patterns with same arity are available,
+    // we follow scalac and silently use the Product one:
+
+    val c1 = C1("s", 0)
+    c1 match {
+      case C1(a, b) =>
+        assert(a == "s")
+        assert(b == 0)
+    }
+
+    // When the size differ, both are patterns become usable:
+
+    val c2 = C2("a", "b")
+    c2 match {
+      case C2(a, b) =>
+        assert(a == "a")
+        assert(b == "b")
+    }
+
+    c2 match {
+      case C2(a, b, c) =>
```

Comment:
That's unchanged. (the first commits come from #1938 which this PR is rebased on to be able to test case class with size > 23)